File: JavaScript/NextGreaterElementII/NextGreaterElementII.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
// (259ms)
var nextGreaterElements = function(nums) {
    stack = [];
    ans = Array(nums.length).fill(-1);
    for(let i = 0; i < nums.length * 2; i++) {
        while(Array.length > 0 && nums[stack[stack.length - 1]] < nums[i % nums.length]) {
            ans[stack.pop()] = nums[i % nums.length];
        }
        if(i < nums.length) {
            stack.push(i);
        }
    }
    return ans;
};

File: Python/NextGreaterElementII/NextGreaterElementII.py
class Solution(object):
    # (339ms)
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        stack = []
        ans = [-1] * len(nums)
        for i in xrange(2 * len(nums)):
            while stack and nums[stack[-1]] < nums[i % len(nums)]:
                ans[stack.pop()] = nums[i % len(nums)]
            if i < len(nums):
                stack.append(i)
        return ans

test = Solution()
print test.nextGreaterElements([1, 2, 1])
print test.nextGreaterElements([5, 4, 3, 2, 1])
